### ISSUE TITLE
Discussion on the use of async in this crate

### ISSUE THREAD

**Issue by MaxVerevkin** [pierrechevalier83/workstyle#20]
There is no need to spawn a dozen of threads. All we need is at most two threads: one for events processing and one for signals (optional).

**Comment by pierrechevalier83** [comment]
(Didn't mean to close the issue before. Fat fingers :sweat_smile: )

**Comment by MaxVerevkin** [comment]
(I can send a RP that removes async stuff if I've convinced you :) )

**Comment by pierrechevalier83** [comment]
> In addition, right now this project depends on both tokio and async-std (via async-i3ipc), which harms compilation times and binary size (for no good reason).

That is a fair point. I am completely ok with a PR (or I'll do it myself when I have a moment) to use `async-std` instead of `tokio` to reduce compilation time/binary size.

> From my understanding, async is useful when there are more tasks than CPU cores.

I don't find this statement convincing. I see async as a way to express what parallelism is possible logically without having to worry about how many cores the user happens to have. Of course, if you want to gain speed through parallelism, you can argue that async is not buying you this in this codebase; but that's not my primary reason for switching to async here. My motivating factor was to increase expressiveness, and incidentally fix at least one known race condition that I had written by hand when I wrote this code spawning threads manually.

> No, everything works, I'm just a bit concerned about wasted resources.

What is the cost of one thread idling on modern hardware? Especially if no more threads will be spun than number of cores. I would see the CPU usage going up, or maybe the cost of spawning threads constantly while the application is running, or even an increase in RAM usage as valid concerns. I don't see it with spawning a few threads, and possibly "under-using" them throughout the lifetime of the process.

> I can send a RP that removes async stuff if I've convinced you

As you can see, I'm not convinced (could change if you showed me one example of actual harm caused by it). However, I would appreciate a PR to replace tokio with async-std if you would like to contribute it.

**Comment by MaxVerevkin** [comment]
> I see async as a way to express what parallelism is possible logically without having to worry about how many cores the user happens to have

Just to be clear, which concurrent tasks are you toking about in context of this project?

> and incidentally fix at least one known race condition that I had written by hand when I wrote this code spawning threads manually

Can you please point me to that bug?

> Especially if no more threads will be spun than number of cores

That's not true. I have a two cores four threads CPU and this is how much threads I get:

![image](https://user-images.githubusercontent.com/34583604/154697275-39a46f6b-1ecd-455a-b33d-0165a47bc5a2.png)

> If so, I would consider configuring tokio to spawn less threads with something like this:

The minimum number of threads we can get using async is 3 (with single threaded `tokio`): main, a blocking `tokio` thread and `swayipc`'s thread). Not sure what happens on `i3` when `async-std` activates. In reality, only two are needed: main and signal handler.

---

By the way, if you really want to use async, I think the config file (that is read on every update, which is on itself is a bug, I suppose) should be read asynchronously too (even though it doesn't actually change anything since there is only one working task anyway). Also there should be `tokio::time::sleep()` instead of `std::thread::sleep()` in the main loop.

**Comment by pierrechevalier83** [comment]
> Just to be clear, which concurrent tasks are you toking about in context of this project?
* register an "event" (something like a window being spawned or killed, so the workspaces should be renamed)
* communicate with the window manager over ipc (establishing connection, get_workspaces, get_tree)
* handling kill signals

> Can you please point me to that bug?
https://github.com/pierrechevalier83/workstyle/issues/15

> That's not true. I have a two cores four threads CPU and this is how much threads I get:
Oh. Looking at it, I also get 30-ish threads on a machine with 16 virtual CPU cores. I assumed tokio would spawn about one thread per core. Maybe the mixing and matching of tokyo and async std causes this.

>  the config file (that is read on every update, which on itself is a bug, I suppose) should be read asynchronously too
You're right about this. That would be a better design. The "bug" isn't really something you can feel as a user as parsing is practically instant; but I agree that it feels hackish and should be refactored. I may do it or would welcome a PR.

> Also there should be tokio::time::sleep() instead of std::thread::sleep() in the main loop.
Good point. Thanks for your feedback.

To sum up the conversation so far, I think these improvements would be welcome:
* switch from tokio to async-std for the executor to be in sync with swayipc (except if https://github.com/JayceFayne/swayipc-rs/pull/34 gets merged upstream, providing an alternative)
* replace std::thread::sleep with the appropriate async equivalent
* read the config file asynchronously on change to it
* consider tuning the executor to spawn less threads (still not sure there is a clear gain, but worth investigating)

I'm happy to take PRs or tackle these feedback points myself, whichever happens first :smile:

Thanks a lot for taking the time to look at my code and share your insights. You're helping me learn and I'm sure the code will get to a better state thanks to this conversation :smile:

**Comment by MaxVerevkin** [comment]
> - register an "event" (something like a window being spawned or killed, so the workspaces should be renamed)
> - communicate with the window manager over ipc (establishing connection, get_workspaces, get_tree)
> - handling kill signals

1\) and 2) are not mutually concurrent, 2) happens _after_ 1). There is no scenario in which you communicate with the wm _while_ waiting for the event to happen. 3) Doesn't interact with the rest of the program - it's purpose is to wait for a signal, drop the lock and call `exit()`.

> Maybe the mixing and matching of tokyo and async std causes this.

No, I'm running `sway` and `async-std` is not involved. `tokio` does spawn one "worker thread" per logial CPU, but it also spawns a number of "blocking threads". Docs:

> These threads are used for blocking operations like tasks spawned through [spawn_blocking](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html). Unlike the [worker_threads](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.worker_threads), they are not always active and will exit if left idle for too long. You can change this timeout duration with [thread_keep_alive](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.thread_keep_alive).
>
> The default value is 512.

> read the config file asynchronously on change to it

Again, isn't this kind of overkill for a simple utility that "just" renames workspace?

**Comment by MaxVerevkin** [comment]
> No, I'm running sway and async-std is not involved

No, turns out async-std implicitly activates :eyes:

**Comment by pierrechevalier83** [comment]
> > No, I'm running sway and async-std is not involved
> 
> No, turns out async-std implicitly activates eyes

I removed tokio in favour of async-std locally yesterday and observed that the same number of workspaces (looks like 2 * number of logical CPU cores + 1, a.k.a 33 in my case) gets spawned.

**Comment by pierrechevalier83** [comment]
Addressed by going back to threads in PR https://github.com/pierrechevalier83/workstyle/pull/26
